sklearn/ensemble/_hist_gradient_boosting/tests/test_gradient_boosting.py

```diff
@@ -172,3 +174,152 @@ def test_binning_train_validation_are_separated():
+        [X_classification, X_regression],
+        [y_classification, y_regression]
+    ):
+        gb = GradientBoosting(max_iter=100, warm_start=True, max_depth=2)
+        gb.fit(X, y)
+        gb.set_params(max_iter=110, max_depth=3)
+        gb.fit(X, y)
+
+        # First 100 trees have max_depth == 2
+        for i in range(100):
+            assert_equal(gb._predictors[i][0].get_max_depth(), 2)
+        # Last 10 trees have max_depth == 3
+        for i in range(1, 11):
+            assert_equal(gb._predictors[-i][0].get_max_depth(), 3)
+
+
+@pytest.mark.parametrize('GradientBoosting, X, y', [
+    (HistGradientBoostingClassifier, X_classification, y_classification),
+    (HistGradientBoostingRegressor, X_regression, y_regression)
+])
+def test_warm_start_early_stopping(GradientBoosting, X, y):
+    # Make sure that early stopping occurs after a small number of iterations
+    # when fitting a second time with warm starting.
+
+    n_iter_no_change = 5
+    gb = GradientBoosting(
+        n_iter_no_change=n_iter_no_change, max_iter=10000,
+        random_state=42, warm_start=True, tol=1e-3
+    )
+    gb.fit(X, y)
+    n_iter_first_fit = gb.n_iter_
+    gb.fit(X, y)
+    n_iter_second_fit = gb.n_iter_
+    assert n_iter_second_fit - n_iter_first_fit < 2 * n_iter_no_change
+
+
+@pytest.mark.parametrize('GradientBoosting, X, y', [
+    (HistGradientBoostingClassifier, X_classification, y_classification),
+    (HistGradientBoostingRegressor, X_regression, y_regression)
+])
+def test_warm_start_equal_n_estimators(GradientBoosting, X, y):
+    # Test if warm start with equal n_estimators does nothing
+    gb_1 = GradientBoosting(max_iter=100, max_depth=2)
+    gb_1.fit(X, y)
+
+    gb_2 = clone(gb_1)
+    gb_2.set_params(max_iter=gb_1.max_iter, warm_start=True)
+    gb_2.fit(X, y)
+
+    # Check identical nodes for each tree
+    for (pred_ith_1, pred_ith_2) in zip(gb_1._predictors, gb_2._predictors):
+        for (predictor_1, predictor_2) in zip(pred_ith_1, pred_ith_2):
+            np.testing.assert_array_equal(
+                predictor_1.nodes,
+                predictor_2.nodes
+            )
+
+    # Check identical predictions
+    np.testing.assert_allclose(gb_1.predict(X), gb_2.predict(X))
+
+
+@pytest.mark.parametrize('GradientBoosting, X, y', [
+    (HistGradientBoostingClassifier, X_classification, y_classification),
+    (HistGradientBoostingRegressor, X_regression, y_regression)
+])
+def test_warm_start_clear(GradientBoosting, X, y):
+    # Test if fit clears state.
+    gb_1 = GradientBoosting(max_iter=100, max_depth=2)
+    gb_1.fit(X, y)
+
+    gb_2 = GradientBoosting(max_iter=100, max_depth=2, warm_start=True)
+    gb_2.fit(X, y)  # inits state
+    gb_2.set_params(warm_start=False)
+    gb_2.fit(X, y)  # clears old state and equals est
+
+    # Check identical nodes for each tree
+    for (pred_ith_1, pred_ith_2) in zip(gb_1._predictors, gb_2._predictors):
```

Comment:
Since we use this pattern a lot, let's make it a small helper `_assert_predictor_equal(gb_1, gb_2)` (in this file, not in `utils`)